crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net_task.rs

```diff
@@ -0,0 +1,1157 @@
+use crate::actor::BroadcastData;
+use crate::event::{
+    FetchOpDataEvt, FetchOpDataEvtQuery, GetAgentInfoSignedEvt, KitsuneP2pEvent,
+    KitsuneP2pEventSender, PutAgentInfoSignedEvt, QueryAgentsEvt,
+};
+use crate::spawn::actor::fetch::FetchResponseConfig;
+use crate::spawn::actor::{
+    Internal, InternalSender, UNAUTHORIZED_DISCONNECT_CODE, UNAUTHORIZED_DISCONNECT_REASON,
+};
+use crate::spawn::meta_net::{
+    nodespace_is_authorized, MetaNetAuth, MetaNetCon, MetaNetEvt, MetaNetEvtRecv, Respond,
+};
+use crate::wire::WireData;
+use crate::{wire, HostApi, KitsuneAgent, KitsuneP2pConfig, KitsuneP2pError, KitsuneSpace};
+use futures::channel::mpsc::Sender;
+use futures::{SinkExt, StreamExt};
+use ghost_actor::GhostSender;
+use kitsune_p2p_fetch::{FetchKey, FetchPool, FetchResponseQueue};
+use kitsune_p2p_timestamp::Timestamp;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+pub struct MetaNetTask {
+    evt_sender: Sender<KitsuneP2pEvent>,
+    host: HostApi,
+    config: KitsuneP2pConfig,
+    fetch_pool: FetchPool,
+    fetch_response_queue: FetchResponseQueue<FetchResponseConfig>,
+    ep_evt: Option<MetaNetEvtRecv>,
+    i_s: GhostSender<Internal>,
+    is_finished: Arc<AtomicBool>,
+}
+
+#[derive(thiserror::Error, Debug)]
+enum MetaNetTaskError {
+    #[error("Ghost actor closed")]
+    GhostActorClosed(#[from] ghost_actor::GhostError),
+
+    #[error("This error should be ignored")]
+    Ignored,
+}
+
+type MetaNetTaskResult<T> = Result<T, MetaNetTaskError>;
+
+impl MetaNetTask {
+    pub fn new(
+        evt_sender: Sender<KitsuneP2pEvent>,
+        host: HostApi,
+        config: KitsuneP2pConfig,
+        fetch_pool: FetchPool,
+        fetch_response_queue: FetchResponseQueue<FetchResponseConfig>,
+        ep_evt: MetaNetEvtRecv,
+        i_s: GhostSender<Internal>,
+    ) -> Self {
+        Self {
+            evt_sender,
+            host,
+            config,
+            fetch_pool,
+            fetch_response_queue,
+            ep_evt: Some(ep_evt),
+            i_s,
+            is_finished: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub fn spawn(mut self) {
+        // Use an mpsc channel rather than a oneshot so no locking is needed in this code to sync the sender.
+        let (shutdown_send, mut shutdown_recv) = futures::channel::mpsc::channel(1);
```

Comment:
There was no shutdown mechanism on fatal errors so I've added this. This is the only functional change I've made though.

Reply:
[Notify](https://docs.rs/tokio/1.32.0/tokio/sync/struct.Notify.html) is lighter weight than a channel if you don't need to send any data.

Reply:
Sounds good, and doesn't require mutability so actually I can simplify this code a bit